Closes #8312:Adds support for ContentDelegate#onExternalResponse

Amejia481 opened the pull request:
I just need to add a new test to `AbstractFetchDownloadService` and some additional sanity check tests, but I would like to get an early feedback.

---
<!-- Text above this line will be added to the commit once "bors" merges this PR -->

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

pocmo commented on components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:
```diff
@@ -727,13 +734,30 @@ class GeckoEngineSession(
+                    destinationDirectory = null,
+                    url = url,
+                    mimeType = contentType
+                )
```
Oh, too bad we do not get those things from GV anymore. Now we have to fish them out of the response ourselves.

Amejia481 replied:
Yeah, there are a lot edge cases that we are going to encounter that were handle by GV. At least, we are finding bugs in our code https://github.com/mozilla-mobile/android-components/issues/8425

pocmo commented on components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:
```diff
@@ -786,6 +810,26 @@ class GeckoEngineSession(
+        } else {
+            null
+        }
+    }
```
I wonder if this is something `guessFileName()` should do whenever we give a mimetype header.

Amejia481 replied:
I guess this is a problem that we are encountering because of reading raw data from headers and in  `guessFileName` is expecting clean `contentType` data. What do you think?

pocmo replied:
Hm, okay. It also takes `contentDisposition` which is a header, maybe the other parameter should also take be able to handle the raw header? 🤔

Amejia481 replied:
I updated the pr and added as part of `guessFileName()`, we will need to do a follow up for it https://github.com/mozilla-mobile/android-components/issues/8425

pocmo commented on components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:
```diff
@@ -786,6 +810,26 @@ class GeckoEngineSession(
+    @VisibleForTesting
+    internal fun WebMessage.getHeaderValue(key: String): String? {
+        return headers[key]?.trim() ?: headers[key.toLowerCase(Locale.ROOT)]?.trim()
+    }
```
I'm confused by this, why do we need to try this again with lowercase?

Amejia481 replied:
I found cases where headers where with capital letters like `Content-Disposition` and others with lowercase `content-disposition`. I think we can just put both in lowercase but it will required that we generate another header object with all the keys in lowercase.

pocmo replied:
@snorp Do you have any advice how Gecko handles this?

snorp replied:
Gecko seems to avoid putting the headers into a map, so it doesn't have this exact problem. They use `nsIHttpChannel::getResponseHeader()`, which performs case-insensitive lookup. We should probably do something similar in `WebMessage`.

Amejia481 replied:
@pocmo  on the new updated gv, `headers`propperty has changed its implementation details to be a`TreeMap` [1] to avoid this issue. Should we remove our code as they are already doing it for us ? But the [public API is still a Map ](https://mozilla.github.io/geckoview/javadoc/mozilla-central/index.html) the change was only on the [implementation details level](https://hg.mozilla.org/mozilla-central/diff/0b410b21eb37f125158648d392713546fc869109/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/WebMessage.java#l1.27).

[1] `Replaced android.util.ArrayMap with java.util.TreeMap in WebMessage to enable case-insensitive handling of the HTTP headers. (bug 1666013)`

pocmo replied:
Yeah, I think I am in favor of removing this code on our side.

Amejia481 replied:
👍

pocmo replied:
After all the behavior is in the public docs of GV now. :)

pocmo commented on components/browser/session/build.gradle:
```diff
@@ -45,6 +45,7 @@ dependencies {
     // We expose this as API because we are using Observable in our public API and do not want every
     // consumer to have to manually import "support-base".
     api project(':support-base')
+    api project(':concept-fetch')
```
Why does `browser-session` need this?

Amejia481 replied:
Thanks for catching this, this was by mistake.

pocmo commented on components/browser/session/src/test/java/mozilla/components/browser/session/engine/EngineObserverTest.kt:
```diff
@@ -700,7 +701,7 @@ class EngineObserverTest {
     fun `onExternalResource will update the store`() {
         val store = BrowserStore()
-
+        val response = mock< Response>()
```
nit: space `mock< Response`

pocmo commented on components/browser/state/src/main/java/mozilla/components/browser/state/state/content/DownloadState.kt:
```diff
@@ -42,7 +45,8 @@ data class DownloadState(
     val sessionId: String? = null,
-    val createdTime: Long = System.currentTimeMillis()
+    val createdTime: Long = System.currentTimeMillis(),
+    val response: @RawValue Response? = null
```
Ugh, that `Parcelize` is scaray here. Do we really need to pass this along as `Parcelable` anywhere?

Amejia481 replied:
We were using it before, but now we don't required it. We could remove support for it.

pocmo replied:
That would be great, I think. :)

Amejia481 commented:
@pocmo we found that new GV api doesn't provide enough info for us to determine the file name for `blob` urls  I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1666013 for it. I'm not sure if we should proceed with landing this pr without it as if we do we are going to lose the functionality for downloading blob urls files. Files are going to be downloaded correctly but we are going to end up with file names like `fileName.bin`.

Amejia481 commented:
> @pocmo we found that new GV api doesn't provide enough info for us to determine the file name for `blob` urls I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1666013 for it. I'm not sure if we should proceed with landing this pr without it as if we do we are going to lose the functionality for downloading blob urls files. Files are going to be downloaded correctly but we are going to end up with file names like `fileName.bin`. This will could fixed many of the downloads bug that we have right now, but it could regress in other cases.

@liuche what do you think? Additionally it will be good to have a bigger audience to test and discover early bugs.

pocmo commented:
I guess we have some time here. The old API is still around. Let's see what the GV team replies in the bug?

Amejia481 commented:
I updated the pr, addressed some of the suggestions + added some missing tests. I'm waiting for an updates on the gv ticket to to see if we should address the remaining suggestions here or if they are going to be addressed on GV.

pocmo commented:
GeckoView bug was fixed :)

Amejia481 commented:
Let's bring this in 🚀 !

Amejia481 commented:
The pr is ready for another review :)

pocmo commented on docs/changelog.md:
```diff
@@ -12,6 +12,8 @@ permalink: /changelog/
 * [Configuration](https://github.com/mozilla-mobile/android-components/blob/master/.config.yml)
 
+* **browser-state**
+  * ⚠️ **This is a breaking change**: Renames `DownloadState` doesn't support [Parcelable](https://developer.android.com/reference/android/os/Parcelable) anymore.
```
"Renames"?

Amejia481 replied:
A residue of a copy paste 🤦 . Fixing it :)

Amejia481 commented on components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:
```diff
@@ -744,15 +750,31 @@ class GeckoEngineSession(
+                val contentType = headers[CONTENT_TYPE]?.trim()
+                val contentLength = headers[CONTENT_LENGTH]?.trim()?.toLong()
+                val contentDisposition = headers[CONTENT_DISPOSITION]?.trim()
+                val url = webResponse.uri
```
Thanks, I missed that one. Yeah, agree it make difficult to identify where variables come from, but at the same time it saves some repetition by not adding the same variable over and over :)

Amejia481 commented:
I filed https://github.com/mozilla-mobile/android-components/issues/8609 for the intermittent test
